core/src/test/scala/integration/kafka/api/PlaintextConsumerCallbackTest.scala

```diff
@@ -0,0 +1,138 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package integration.kafka.api
+
+import kafka.api.{AbstractConsumerTest, BaseConsumerTest}
+import kafka.utils.{TestInfoUtils, TestUtils}
+import org.apache.kafka.clients.consumer.{Consumer, ConsumerRebalanceListener}
+import org.apache.kafka.common.TopicPartition
+import org.junit.jupiter.api.Assertions.{assertEquals, assertThrows, assertTrue}
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.{Arguments, MethodSource}
+
+import java.util
+import java.util.Arrays.asList
+import java.util.Collections
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.stream.Stream
+
+/**
+ * Integration tests for the consumer that cover interaction with the consumer from within callbacks
+ * and listeners.
+ */
+class PlaintextConsumerCallbackTest extends AbstractConsumerTest {
+
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
+  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersAll"))
+  def testConsumerRebalanceListenerAssignOnPartitionsAssigned(quorum: String, groupProtocol: String): Unit = {
+    val tp = new TopicPartition(topic, 0);
+    triggerOnPartitionsAssigned { (consumer, _) =>
+      val e: Exception = assertThrows(classOf[IllegalStateException], () => consumer.assign(Collections.singletonList(tp)))
+      assertEquals(e.getMessage, "Subscription to topics, partitions and pattern are mutually exclusive")
```

Comment:
nit: the message is totally accurate but this test has nothing to do with pattern subscription, so maybe clearer for whoever gets the failure to be specific about the problem on this test (subscribe & assign mutually exclusive). We'll probably end up adding another test where the failure would be calling subscribe(Pattern) on the callback, and that one would require the message for the 2 subscribe calls being mutually exclusive.

Reply:
Are you proposing to change the message of the exception that is verified here?

Reply:
nop, just adding a message to the assertThrows so that the failure comes out specific to calling assign after subscribe (and then we would end up with a similar assertThrows for calling subscribe(Pattern)), but re-thinking it, it's just clear enough in case of failure since it would most probably always fail in the call to assign/subscribePattern itself. No concerns then.